File: samlang-checker/src/main/kotlin/samlang/checker/TypeValidator.kt
package samlang.checker

import samlang.ast.common.Range
import samlang.ast.common.Type
import samlang.ast.common.Type.FunctionType
import samlang.ast.common.Type.IdentifierType
import samlang.ast.common.Type.PrimitiveType
import samlang.ast.common.Type.TupleType
import samlang.ast.common.Type.UndecidedType
import samlang.ast.common.TypeVisitor
import samlang.errors.NotWellDefinedIdentifierError

internal object TypeValidator {
    fun validateType(
        type: Type,
        identifierTypeValidator: IdentifierTypeValidator,
        errorCollector: ErrorCollector,
        errorRange: Range
    ): Boolean {
        val badIdentifier = validateType(type = type, identifierTypeValidator = identifierTypeValidator)
            ?: return true
        errorCollector.add(NotWellDefinedIdentifierError(badIdentifier = badIdentifier, range = errorRange))
        return false
    }

    /** Exposed for test */
    fun validateType(type: Type, identifierTypeValidator: IdentifierTypeValidator): String? {
        val visitor = Visitor()
        type.accept(visitor = visitor, context = identifierTypeValidator)
        return visitor.badIdentifier
    }

    /** A validator for type to check whether every identifier type is well defined.*/
    private class Visitor : TypeVisitor<IdentifierTypeValidator, Boolean> {
        var badIdentifier: String? = null

        override fun visit(type: PrimitiveType, context: IdentifierTypeValidator): Boolean = true

        override fun visit(type: IdentifierType, context: IdentifierTypeValidator): Boolean {
            val (name, typeArguments) = type
            if (!context.identifierTypeIsWellDefined(name = name, typeArgumentLength = typeArguments.size)) {
                badIdentifier = name
                return false
            }
            return validate(list = typeArguments, context = context)
        }

        override fun visit(type: TupleType, context: IdentifierTypeValidator): Boolean =
            validate(list = type.mappings, context = context)

        override fun visit(type: FunctionType, context: IdentifierTypeValidator): Boolean =
            validate(list = type.argumentTypes, context = context) &&
                    type.returnType.accept(visitor = this, context = context)

        override fun visit(type: UndecidedType, context: IdentifierTypeValidator): Boolean = true

        private fun validate(list: List<Type>, context: IdentifierTypeValidator): Boolean =
            list.all { it.accept(visitor = this, context = context) }
    }
}

File: samlang-core-checker/type-validator.ts
import type { Type, Range } from 'samlang-core-ast/common-nodes';
import type { ModuleErrorCollector } from 'samlang-core-errors';
import { checkNotNull } from 'samlang-core-utils';

export interface IdentifierTypeValidator {
  /**
   * Given the `name` of the identifier and the number of applied arguments, check the context to see whether it
   * matches any type definition in scope.
   */
  identifierTypeIsWellDefined(name: string, typeArgumentLength: number): boolean;
}

/**
 * @returns an invalidid string identifier, null if there is none.
 */
// eslint-disable-next-line camelcase
export const findInvalidTypeIdentifier_EXPOSED_FOR_TESTING = (
  type: Type,
  identifierTypeValidator: IdentifierTypeValidator
): string | null => {
  switch (type.type) {
    case 'PrimitiveType':
      return null;
    case 'IdentifierType':
      if (
        !identifierTypeValidator.identifierTypeIsWellDefined(
          type.identifier,
          type.typeArguments.length
        )
      ) {
        return type.identifier;
      }
      return findInvalidTypeIdentifierForList(type.typeArguments, identifierTypeValidator);
    case 'TupleType':
      return findInvalidTypeIdentifierForList(type.mappings, identifierTypeValidator);
    case 'FunctionType':
      return (
        findInvalidTypeIdentifier_EXPOSED_FOR_TESTING(type.returnType, identifierTypeValidator) ??
        findInvalidTypeIdentifierForList(type.argumentTypes, identifierTypeValidator)
      );
    case 'UndecidedType':
      return null;
  }
};

const findInvalidTypeIdentifierForList = (
  types: readonly Type[],
  identifierTypeValidator: IdentifierTypeValidator
): string | null => {
  for (let i = 0; i < types.length; i += 1) {
    const invalidName = findInvalidTypeIdentifier_EXPOSED_FOR_TESTING(
      checkNotNull(types[i]),
      identifierTypeValidator
    );
    if (invalidName != null) {
      return invalidName;
    }
  }
  return null;
};

export const validateType = (
  type: Type,
  identifierTypeValidator: IdentifierTypeValidator,
  errorCollector: ModuleErrorCollector,
  errorRange: Range
): boolean => {
  const badIdentifier = findInvalidTypeIdentifier_EXPOSED_FOR_TESTING(
    type,
    identifierTypeValidator
  );
  if (badIdentifier == null) {
    return true;
  }
  errorCollector.reportNotWellDefinedIdentifierError(errorRange, badIdentifier);
  return false;
};
